# **Segmenting and Clustering Neighborhoods in Toronto**


### 1. Build Dataframe

In [1]:
import pandas as pd
import html5lib
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
tables = pd.read_html(url)
table = tables[0]

In [2]:
table.head()

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

In [3]:
table.tail()

Postcode       Borough          Neighbourhood
282      M8Z     Etobicoke              Mimico NW
283      M8Z     Etobicoke     The Queensway West
284      M8Z     Etobicoke  Royal York South West
285      M8Z     Etobicoke         South of Bloor
286      M9Z  Not assigned           Not assigned

In [4]:
table.shape

(287, 3)

In [5]:
table.describe(include='all')

Postcode       Borough Neighbourhood
count       287           287           287
unique      180            12           208
top         M8Y  Not assigned  Not assigned
freq          8            77            78


 As being asked, we will removethe cells with a borough*** that is 'Not assigned'.

In [6]:
table_temp1=table.loc[table['Borough']!='Not assigned'].reset_index(drop=True)
table_temp1

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M6A        North York   
4        M6A        North York   
5        M7A      Queen's Park   
6        M9A      Queen's Park   
7        M1B       Scarborough   
8        M1B       Scarborough   
9        M3B        North York   
10       M4B         East York   
11       M4B         East York   
12       M5B  Downtown Toronto   
13       M5B  Downtown Toronto   
14       M6B        North York   
15       M9B         Etobicoke   
16       M9B         Etobicoke   
17       M9B         Etobicoke   
18       M9B         Etobicoke   
19       M9B         Etobicoke   
20       M1C       Scarborough   
21       M1C       Scarborough   
22       M1C       Scarborough   
23       M3C        North York   
24       M3C        North York   
25       M4C         East York   
26       M5C  Downtown Toronto   
27       M6C              York   
28       M9C         Etobicoke   
29       M9C         Etobicoke   
..       ...               ...   
180      M9V         Etobicoke   
181      M1W       Scarborough   
182      M4W  Downtown Toronto   
183      M5W  Downtown Toronto   
184      M8W         Etobicoke   
185      M8W         Etobicoke   
186      M9W         Etobicoke   
187      M1X       Scarborough   
188      M4X  Downtown Toronto   
189      M4X  Downtown Toronto   
190      M5X  Downtown Toronto   
191      M5X  Downtown Toronto   
192      M8X         Etobicoke   
193      M8X         Etobicoke   
194      M8X         Etobicoke   
195      M4Y  Downtown Toronto   
196      M7Y      East Toronto   
197      M8Y         Etobicoke   
198      M8Y         Etobicoke   
199      M8Y         Etobicoke   
200      M8Y         Etobicoke   
201      M8Y         Etobicoke   
202      M8Y         Etobicoke   
203      M8Y         Etobicoke   
204      M8Y         Etobicoke   
205      M8Z         Etobicoke   
206      M8Z         Etobicoke   
207      M8Z         Etobicoke   
208      M8Z         Etobicoke   
209      M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                                         Harbourfront  
3                                     Lawrence Heights  
4                                       Lawrence Manor  
5                                         Not assigned  
6                                         Queen's Park  
7                                                Rouge  
8                                              Malvern  
9                                      Don Mills North  
10                                    Woodbine Gardens  
11                                       Parkview Hill  
12                                             Ryerson  
13                                     Garden District  
14                                           Glencairn  
15                                          Cloverdale  
16                                           Islington  
17                                        Martin Grove  
18                                    Princess Gardens  
19                                     West Deane Park  
20                                      Highland Creek  
21                                          Rouge Hill  
22                                          Port Union  
23                                     Flemingdon Park  
24                                     Don Mills South  
25                                    Woodbine Heights  
26                                      St. James Town  
27                                  Humewood-Cedarvale  
28                                   Bloordale Gardens  
29                                            Eringate  
..                                                 ...  
180                                        Thistletown  
181           

Then, for neighborhoods having the same postal code, **combine** them separated with a comma.

In [7]:
table_temp2=table_temp1.groupby('Postcode',as_index=False).agg(lambda x:','.join(set(x)))
table_temp2.tail(20)

Postcode       Borough                                      Neighbourhood
83       M6R  West Toronto                              Roncesvalles,Parkdale
84       M6S  West Toronto                                  Runnymede,Swansea
85       M7A  Queen's Park                                       Not assigned
86       M7R   Mississauga              Canada Post Gateway Processing Centre
87       M7Y  East Toronto  Business Reply Mail Processing Centre 969 Eastern
88       M8V     Etobicoke         New Toronto,Humber Bay Shores,Mimico South
89       M8W     Etobicoke                              Long Branch,Alderwood
90       M8X     Etobicoke        The Kingsway,Old Mill North,Montgomery Road
91       M8Y     Etobicoke  The Queensway East,Humber Bay,Sunnylea,Royal Y...
92       M8Z     Etobicoke  Kingsway Park South West,Mimico NW,The Queensw...
93       M9A  Queen's Park                                       Queen's Park
94       M9B     Etobicoke  Cloverdale,Islington,West Deane Park,Princess ...
95       M9C     Etobicoke  Markland Wood,Old Burnhamthorpe,Bloordale Gard...
96       M9L    North York                                      Humber Summit
97       M9M    North York                                    Emery,Humberlea
98       M9N          York                                             Weston
99       M9P     Etobicoke                                          Westmount
100      M9R     Etobicoke  Richview Gardens,Martin Grove Gardens,Kingsvie...
101      M9V     Etobicoke  Beaumond Heights,Jamestown,Mount Olive,Albion ...
102      M9W     Etobicoke                                          Northwest

Now, as the assignment asked, if a cell has a borough but a Not assigned neighborhood, then the neighborhood will be given the same as the borough.

In [8]:
table_temp2.loc[table_temp2['Neighbourhood']=='Not assigned','Neighbourhood']=table_temp2['Borough']

In [9]:
table_temp2.tail(20)

Postcode       Borough                                      Neighbourhood
83       M6R  West Toronto                              Roncesvalles,Parkdale
84       M6S  West Toronto                                  Runnymede,Swansea
85       M7A  Queen's Park                                       Queen's Park
86       M7R   Mississauga              Canada Post Gateway Processing Centre
87       M7Y  East Toronto  Business Reply Mail Processing Centre 969 Eastern
88       M8V     Etobicoke         New Toronto,Humber Bay Shores,Mimico South
89       M8W     Etobicoke                              Long Branch,Alderwood
90       M8X     Etobicoke        The Kingsway,Old Mill North,Montgomery Road
91       M8Y     Etobicoke  The Queensway East,Humber Bay,Sunnylea,Royal Y...
92       M8Z     Etobicoke  Kingsway Park South West,Mimico NW,The Queensw...
93       M9A  Queen's Park                                       Queen's Park
94       M9B     Etobicoke  Cloverdale,Islington,West Deane Park,Princess ...
95       M9C     Etobicoke  Markland Wood,Old Burnhamthorpe,Bloordale Gard...
96       M9L    North York                                      Humber Summit
97       M9M    North York                                    Emery,Humberlea
98       M9N          York                                             Weston
99       M9P     Etobicoke                                          Westmount
100      M9R     Etobicoke  Richview Gardens,Martin Grove Gardens,Kingsvie...
101      M9V     Etobicoke  Beaumond Heights,Jamestown,Mount Olive,Albion ...
102      M9W     Etobicoke                                          Northwest

Finally, let's check the shape of the dataframe as required. **Rename the columns** as the assignment used for consistency.

In [10]:
table_cy=table_temp2
table_cy.rename(columns={'Postcode':'PostalCode','Neighbourhood':'Neighborhood'},inplace=True)
table_cy.head()

PostalCode      Borough                          Neighborhood
0        M1B  Scarborough                         Rouge,Malvern
1        M1C  Scarborough  Port Union,Highland Creek,Rouge Hill
2        M1E  Scarborough       West Hill,Guildwood,Morningside
3        M1G  Scarborough                                Woburn
4        M1H  Scarborough                             Cedarbrae

In [11]:
table_cy.shape

(103, 3)

So now it has 103 rows.

### 2.Get Latitude/Longitude Coordinates

In [12]:
csv='https://cocl.us/Geospatial_data'
Geospatial=pd.read_csv(csv)
Geospatial.rename(columns={'Postal Code':'PostalCode'},inplace=True)
Geospatial

PostalCode   Latitude  Longitude
0          M1B  43.806686 -79.194353
1          M1C  43.784535 -79.160497
2          M1E  43.763573 -79.188711
3          M1G  43.770992 -79.216917
4          M1H  43.773136 -79.239476
5          M1J  43.744734 -79.239476
6          M1K  43.727929 -79.262029
7          M1L  43.711112 -79.284577
8          M1M  43.716316 -79.239476
9          M1N  43.692657 -79.264848
10         M1P  43.757410 -79.273304
11         M1R  43.750072 -79.295849
12         M1S  43.794200 -79.262029
13         M1T  43.781638 -79.304302
14         M1V  43.815252 -79.284577
15         M1W  43.799525 -79.318389
16         M1X  43.836125 -79.205636
17         M2H  43.803762 -79.363452
18         M2J  43.778517 -79.346556
19         M2K  43.786947 -79.385975
20         M2L  43.757490 -79.374714
21         M2M  43.789053 -79.408493
22         M2N  43.770120 -79.408493
23         M2P  43.752758 -79.400049
24         M2R  43.782736 -79.442259
25         M3A  43.753259 -79.329656
26         M3B  43.745906 -79.352188
27         M3C  43.725900 -79.340923
28         M3H  43.754328 -79.442259
29         M3J  43.767980 -79.487262
..         ...        ...        ...
73         M6C  43.693781 -79.428191
74         M6E  43.689026 -79.453512
75         M6G  43.669542 -79.422564
76         M6H  43.669005 -79.442259
77         M6J  43.647927 -79.419750
78         M6K  43.636847 -79.428191
79         M6L  43.713756 -79.490074
80         M6M  43.691116 -79.476013
81         M6N  43.673185 -79.487262
82         M6P  43.661608 -79.464763
83         M6R  43.648960 -79.456325
84         M6S  43.651571 -79.484450
85         M7A  43.662301 -79.389494
86         M7R  43.636966 -79.615819
87         M7Y  43.662744 -79.321558
88         M8V  43.605647 -79.501321
89         M8W  43.602414 -79.543484
90         M8X  43.653654 -79.506944
91         M8Y  43.636258 -79.498509
92         M8Z  43.628841 -79.520999
93         M9A  43.667856 -79.532242
94         M9B  43.650943 -79.554724
95         M9C  43.643515 -79.577201
96         M9L  43.756303 -79.565963
97         M9M  43.724766 -79.532242
98         M9N  43.706876 -79.518188
99         M9P  43.696319 -79.532242
100        M9R  43.688905 -79.554724
101        M9V  43.739416 -79.588437
102        M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [13]:
table_use=pd.merge(table_cy,Geospatial,left_on='PostalCode',right_on='PostalCode')
table_use

PostalCode           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
..         ...               ...   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A      Queen's Park   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A      Queen's Park   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                        Rouge,Malvern  43.806686 -79.194353  
1                 Port Union,Highland Creek,Rouge Hill  43.784535 -79.160497  
2                      West Hill,Guildwood,Morningside  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
5                                  Scarborough Village  43.744734 -79.239476  
6            Kennedy Park,East Birchmount Park,Ionview  43.727929 -79.262029  
7                        Clairlea,Golden Mile,Oakridge  43.711112 -79.284577  
8        Cliffside,Scarborough Village West,Cliffcrest  43.716316 -79.239476  
9                           Birch Cliff,Cliffside West  43.692657 -79.264848  
10   Dorset Park,Wexford Heights,Scarborough Town C...  43.757410 -79.273304  
11                                    Wexford,Maryvale  43.750072 -79.295849  
12                                           Agincourt  43.794200 -79.262029  
13               Clarks Corners,Tam O'Shanter,Sullivan  43.781638 -79.304302  
14   Agincourt North,L'Amoreaux East,Steeles East,M...  43.815252 -79.284577  
15                                     L'Amoreaux West  43.799525 -79.318389  
16                                         Upper Rouge  43.836125 -79.205636  
17                                   Hillcrest Village  43.803762 -79.363452  
18                          Henry Farm,Oriole,Fairview  43.778517 -79.346556  
19                                     Bayview Village  43.786947 -79.385975  
20                             York Mills,Silver Hills  43.757490 -79.374714  
21                              N

In [14]:
table_use.shape

(103, 5)

In [15]:
table_use.describe(include='all')

PostalCode     Borough  Neighborhood    Latitude   Longitude
count         103         103           103  103.000000  103.000000
unique        103          11           102         NaN         NaN
top           M5W  North York  Queen's Park         NaN         NaN
freq            1          24             2         NaN         NaN
mean          NaN         NaN           NaN   43.704608  -79.397153
std           NaN         NaN           NaN    0.052463    0.097146
min           NaN         NaN           NaN   43.602414  -79.615819
25%           NaN         NaN           NaN   43.660567  -79.464763
50%           NaN         NaN           NaN   43.696948  -79.388790
75%           NaN         NaN           NaN   43.745320  -79.340923
max           NaN         NaN           NaN   43.836125  -79.160497

### 3. Explore and cluster the neighborhoods in Toronto

Before we start exploring, let's download key libraries that we may need to use.

In [16]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library


Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    ca-certificates-2019.9.11  |       hecc5488_0         144 KB  conda-forge
    certifi-2019.9.11          |           py36_0         147 KB  conda-forge
    geopy-1.20.0               |             py_0          57 KB  conda-forge
    openssl-1.1.1d             |       h516909a_0         2.1 MB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0        conda-forge
    geopy:           1.20.0-py_0      conda-forge

The following packages will be UPDATED:

    certi

In [17]:
table_use.Borough.value_counts()

North York          24
Downtown Toronto    18
Scarborough         17
Etobicoke           11
Central Toronto      9
West Toronto         6
East Toronto         5
York                 5
East York            5
Queen's Park         2
Mississauga          1
Name: Borough, dtype: int64

#### **As asked, now let's select those boroughs that contain the word Toronto.** 

In [18]:
table_tor=table_use[table_use.Borough.str.contains('Toronto')].reset_index(drop=True)
table_tor.head()

PostalCode          Borough                   Neighborhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    Riverdale,The Danforth West  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [19]:
table_tor

PostalCode           Borough  \
0         M4E      East Toronto   
1         M4K      East Toronto   
2         M4L      East Toronto   
3         M4M      East Toronto   
4         M4N   Central Toronto   
5         M4P   Central Toronto   
6         M4R   Central Toronto   
7         M4S   Central Toronto   
8         M4T   Central Toronto   
9         M4V   Central Toronto   
10        M4W  Downtown Toronto   
11        M4X  Downtown Toronto   
12        M4Y  Downtown Toronto   
13        M5A  Downtown Toronto   
14        M5B  Downtown Toronto   
15        M5C  Downtown Toronto   
16        M5E  Downtown Toronto   
17        M5G  Downtown Toronto   
18        M5H  Downtown Toronto   
19        M5J  Downtown Toronto   
20        M5K  Downtown Toronto   
21        M5L  Downtown Toronto   
22        M5N   Central Toronto   
23        M5P   Central Toronto   
24        M5R   Central Toronto   
25        M5S  Downtown Toronto   
26        M5T  Downtown Toronto   
27        M5V  Downtown Toronto   
28        M5W  Downtown Toronto   
29        M5X  Downtown Toronto   
30        M6G  Downtown Toronto   
31        M6H      West Toronto   
32        M6J      West Toronto   
33        M6K      West Toronto   
34        M6P      West Toronto   
35        M6R      West Toronto   
36        M6S      West Toronto   
37        M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                                         The Beaches  43.676357 -79.293031  
1                         Riverdale,The Danforth West  43.679557 -79.352188  
2                       The Beaches West,India Bazaar  43.668999 -79.315572  
3                                     Studio District  43.659526 -79.340923  
4                                       Lawrence Park  43.728020 -79.388790  
5                                    Davisville North  43.712751 -79.390197  
6                                  North Toronto West  43.715383 -79.405678  
7                                          Davisville  43.704324 -79.388790  
8                          Moore Park,Summerhill East  43.689574 -79.383160  
9   Rathnelly,Forest Hill SE,Summerhill West,Deer ...  43.686412 -79.400049  
10                                           Rosedale  43.679563 -79.377529  
11                         Cabbagetown,St. James Town  43.667967 -79.367675  
12                               Church and Wellesley  43.665860 -79.383160  
13                                       Harbourfront  43.654260 -79.360636  
14                            Garden District,Ryerson  43.657162 -79.378937  
15                                     St. James Town  43.651494 -79.375418  
16                                        Berczy Park  43.644771 -79.373306  
17                                 Central Bay Street  43.657952 -79.387383  
18                             Adelaide,Richmond,King  43.650571 -79.384568  
19    Toronto Islands,Harbourfront East,Union Station  43.640816 -79.381752  
20            Toronto Dominion Centre,Design Exchange  43.647177 -79.381576  
21                      Victoria Hotel,Commerce Court  43.648198 -79.379817  
22                                           Roselawn  43.711695 -79.416936  
23                 Forest Hill North,Forest Hill West  43.696948 -79.411307  
24                  North Midtown,Yorkville,The Annex  43.672710 -79.405678  
25                      Harbord,University of Toronto  43.662696 -79.400049  
26            Grange Park,Kensington Market,Chinatown  43.653206 -79.400049  
27  King and Spadina,Harbourfront West,Bathurst Qu...  43.628947 -79.394420  
28                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846  
29              Underground city,First Canadian Place  43.648429 -79.382280  
30                                           Christie  43.669542 -79.422564  
31                        Dufferin,Dovercourt Village  43.669005 -79.442259  
32                            Little Portugal,Trinity  43.64792

In [20]:
table_tor.shape

(38, 5)

In [21]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


In [22]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, neigh, borough in zip(table_tor['Latitude'], table_tor['Longitude'], table_tor['Neighborhood'], table_tor['Borough']):
    label = '{}, {}'.format(neigh, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Next Define Foursquare Credentials and Version

In [23]:
CLIENT_ID = 'CLEBLHUA2WPRPAEWW5X4U5IPJ23GG2JPUZT34N0EKX54HVZQ' # your Foursquare ID
CLIENT_SECRET = 'RCCHJNMELZ5VGBOH4M3HXQMGDF2OPY24JXLS1JZ0M25ACXEL' # your Foursquare Secret
VERSION = '20190916' # Foursquare API version

print('My credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

My credentails:
CLIENT_ID: CLEBLHUA2WPRPAEWW5X4U5IPJ23GG2JPUZT34N0EKX54HVZQ
CLIENT_SECRET:RCCHJNMELZ5VGBOH4M3HXQMGDF2OPY24JXLS1JZ0M25ACXEL


#### **As a test, let's check the top 100 venues that are in a downtown neighborhood within a radius of 500 meters.**

In [24]:
table_tor.loc[20, 'Neighborhood']

'Toronto Dominion Centre,Design Exchange'

In [25]:
neighborhood_latitude = table_tor.loc[20, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = table_tor.loc[20, 'Longitude'] # neighborhood longitude value

neighborhood_name = table_tor.loc[20, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Toronto Dominion Centre,Design Exchange are 43.6471768, -79.38157640000001.


In [26]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL


'https://api.foursquare.com/v2/venues/explore?&client_id=CLEBLHUA2WPRPAEWW5X4U5IPJ23GG2JPUZT34N0EKX54HVZQ&client_secret=RCCHJNMELZ5VGBOH4M3HXQMGDF2OPY24JXLS1JZ0M25ACXEL&v=20190916&ll=43.6471768,-79.38157640000001&radius=500&limit=100'

In [27]:
results = requests.get(url).json()
#results

From the Foursquare lab in the previous module, we know that all the information is in the ***items*** key. Before we proceed, let's borrow the **get_category_type** function from the Foursquare lab.

In [28]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a *pandas* dataframe.

Get the ***items*** key.

In [29]:
venues = results['response']['groups'][0]['items']
#venues

In [30]:
nearby_venues = json_normalize(venues) # flatten JSON
nearby_venues.head(2)

reasons.count                                      reasons.items  \
0              0  [{'summary': 'This spot is popular', 'type': '...   
1              0  [{'summary': 'This spot is popular', 'type': '...   

                       referralId  \
0  e-0-501ae947e4b0d11883b910a7-0   
1  e-0-4ad4c05df964a52059f620e3-1   

                                    venue.categories  \
0  [{'id': '4bf58dd8d48988d176941735', 'name': 'G...   
1  [{'id': '4bf58dd8d48988d1c4941735', 'name': 'R...   

                   venue.id venue.location.address venue.location.cc  \
0  501ae947e4b0d11883b910a7             199 Bay St                CA   
1  4ad4c05df964a52059f620e3  66 Wellington St West                CA   

  venue.location.city venue.location.country  \
0             Toronto                 Canada   
1             Toronto                 Canada   

           venue.location.crossStreet  venue.location.distance  \
0  at Commerce Court West, PATH Level                      164   
1                       at Bay Street                       36   

                     venue.location.formattedAddress  \
0  [199 Bay St (at Commerce Court West, PATH Leve...   
1  [66 Wellington St West (at Bay Street), Toront...   

                       venue.location.labeledLatLngs  venue.location.lat  \
0  [{'label': 'display', 'lat': 43.64809974034856...           43.648100   
1  [{'label': 'display', 'lat': 43.64745206618347...           43.647452   

   venue.location.lng venue.location.neighborhood venue.location.postalCode  \
0          -79.379989                         NaN                   M5L 1L5   
1          -79.381320                         NaN                   M5K 1H6   

  venue.location.state          venue.name  venue.photos.count  \
0                   ON  Equinox Bay Street                   0   
1                   ON               Canoe                   0   

  venue.photos.groups venue.venuePage.id  
0                  []                NaN  
1                  []           40552945

In [31]:
# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]
nearby_venues.head(2)

venue.name                                   venue.categories  \
0  Equinox Bay Street  [{'id': '4bf58dd8d48988d176941735', 'name': 'G...   
1               Canoe  [{'id': '4bf58dd8d48988d1c4941735', 'name': 'R...   

   venue.location.lat  venue.location.lng  
0           43.648100          -79.379989  
1           43.647452          -79.381320

In [32]:
# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)
nearby_venues.head(2)

venue.name venue.categories  venue.location.lat  venue.location.lng
0  Equinox Bay Street              Gym           43.648100          -79.379989
1               Canoe       Restaurant           43.647452          -79.381320

In [33]:
# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]
nearby_venues.head(2)

name  categories        lat        lng
0  Equinox Bay Street         Gym  43.648100 -79.379989
1               Canoe  Restaurant  43.647452 -79.381320

In [34]:
nearby_venues.shape

(100, 4)

#### **Now let's repeat the venue checking process for all the neighborhoods.**

In [35]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [36]:
toronto_venues=getNearbyVenues(names=table_tor['Neighborhood'],
                                   latitudes=table_tor['Latitude'],
                                   longitudes=table_tor['Longitude']
                                  )

The Beaches
Riverdale,The Danforth West
The Beaches West,India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park,Summerhill East
Rathnelly,Forest Hill SE,Summerhill West,Deer Park,South Hill
Rosedale
Cabbagetown,St. James Town
Church and Wellesley
Harbourfront
Garden District,Ryerson
St. James Town
Berczy Park
Central Bay Street
Adelaide,Richmond,King
Toronto Islands,Harbourfront East,Union Station
Toronto Dominion Centre,Design Exchange
Victoria Hotel,Commerce Court
Roselawn
Forest Hill North,Forest Hill West
North Midtown,Yorkville,The Annex
Harbord,University of Toronto
Grange Park,Kensington Market,Chinatown
King and Spadina,Harbourfront West,Bathurst Quay,Railway Lands,Island airport,South Niagara,CN Tower
Stn A PO Boxes 25 The Esplanade
Underground city,First Canadian Place
Christie
Dufferin,Dovercourt Village
Little Portugal,Trinity
Parkdale Village,Brockton,Exhibition Place
High Park,The Junction South
Roncesvalles,Parkdale
Runnymede

In [37]:
print(toronto_venues.shape)
toronto_venues.iloc[0:1702:200,:]

(1702, 7)


Neighborhood  Neighborhood Latitude  \
0                                    The Beaches              43.676357   
200                   Cabbagetown,St. James Town              43.667967   
400                      Garden District,Ryerson              43.657162   
600                                  Berczy Park              43.644771   
800                       Adelaide,Richmond,King              43.650571   
1000     Toronto Dominion Centre,Design Exchange              43.647177   
1200     Grange Park,Kensington Market,Chinatown              43.653206   
1400       Underground city,First Canadian Place              43.648429   
1600  Parkdale Village,Brockton,Exhibition Place              43.636847   

      Neighborhood Longitude                             Venue  \
0                 -79.293031                 Glen Manor Ravine   
200               -79.367675            Butter Chicken Factory   
400               -79.378937              Ryerson Image Centre   
600               -79.373306                         Starbucks   
800               -79.384568               Bell Trinity Square   
1000              -79.381576                           deKEFIR   
1200              -79.400049  The Dirty Bird Chicken + Waffles   
1400              -79.382280       Sam James Coffee Bar (SJCB)   
1600              -79.428191                 Vogue Supper Club   

      Venue Latitude  Venue Longitude           Venue Category  
0          43.676821       -79.293942                    Trail  
200        43.667072       -79.369184        Indian Restaurant  
400        43.657523       -79.379460              Art Gallery  
600        43.644525       -79.368560              Coffee Shop  
800        43.653475       -79.382470                   Office  
1000       43.650433       -79.379393           Ice Cream Shop  
1200       43.654612       -79.400674  Comfort Food Restaurant  
1400       43.647881       -79.384332                     Café  
1600       43.636951       -79.425446               Restaurant

In [38]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  The Beaches              43.676357              -79.293031   
1  The Beaches              43.676357              -79.293031   
2  The Beaches              43.676357              -79.293031   
3  The Beaches              43.676357              -79.293031   
4  The Beaches              43.676357              -79.293031   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                   Glen Stewart Park       43.675278       -79.294647   
4                       Upper Beaches       43.680563       -79.292869   

      Venue Category  
0              Trail  
1  Health Food Store  
2                Pub  
3               Park  
4       Neighborhood

In [39]:
len(toronto_venues['Venue Category'].unique())

230

In [40]:
#toronto_venues['Venue Category'].value_counts()

In [41]:
#toronto_venues['Venue Category'].unique()

In [42]:
toronto_venues[(toronto_venues['Venue Category'] == 'Neighborhood')]

Neighborhood  Neighborhood Latitude  \
4                                        The Beaches              43.676357   
72                                   Studio District              43.659526   
729                           Adelaide,Richmond,King              43.650571   
816  Toronto Islands,Harbourfront East,Union Station              43.640816   

     Neighborhood Longitude             Venue  Venue Latitude  \
4                -79.293031     Upper Beaches       43.680563   
72               -79.340923       Leslieville       43.662070   
729              -79.384568  Downtown Toronto       43.653232   
816              -79.381752      Harbourfront       43.639526   

     Venue Longitude Venue Category  
4         -79.292869   Neighborhood  
72        -79.337856   Neighborhood  
729       -79.385296   Neighborhood  
816       -79.380688   Neighborhood

In [43]:
toronto_venues.loc[toronto_venues['Venue Category']=='Neighborhood']

Neighborhood  Neighborhood Latitude  \
4                                        The Beaches              43.676357   
72                                   Studio District              43.659526   
729                           Adelaide,Richmond,King              43.650571   
816  Toronto Islands,Harbourfront East,Union Station              43.640816   

     Neighborhood Longitude             Venue  Venue Latitude  \
4                -79.293031     Upper Beaches       43.680563   
72               -79.340923       Leslieville       43.662070   
729              -79.384568  Downtown Toronto       43.653232   
816              -79.381752      Harbourfront       43.639526   

     Venue Longitude Venue Category  
4         -79.292869   Neighborhood  
72        -79.337856   Neighborhood  
729       -79.385296   Neighborhood  
816       -79.380688   Neighborhood

#### **Next let's analyze each neighborhood.**

In [44]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot.head()

Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0             0         0            0                    0                 0   
1             0         0            0                    0                 0   
2             0         0            0                    0                 0   
3             0         0            0                    0                 0   
4             0         0            0                    0                 0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                   0              0          0           0           0   
1                   0              0          0           0           0   
2                   0              0          0           0           0   
3                   0              0          0           0           0   
4                   0              0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       0     0    0                 0                   0      0   
1       0     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0           0              0          0         0                     0   
1           0              0          0         0                     0   
2           0              0          0         0                     0   
3           0              0          0         0                     0   
4           0              0          0         0                     0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Building  Burger Joint  \
0               0        0                0         0             0   
1               0        0                0         0             0   
2               0        0                0         0             0   
3               0        0                0         0             0   
4               0        0                0         0             0   

   Burrito Place  Bus Line  Butcher  Café  Cajun / Creole Restaurant  \
0              0         0        0     0                          0   
1              0         0        0     0                          0   
2              0         0        0     0                          0   
3              0         0        0     0                          0   
4              0         0        0     0                          0   

   Candy Store  Caribbean Restaurant  Ch

In [45]:
# add neighborhood column back to dataframe
toronto_onehot['TorNeighborhood'] = toronto_venues['Neighborhood'] 
toronto_onehot.head()

Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0             0         0            0                    0                 0   
1             0         0            0                    0                 0   
2             0         0            0                    0                 0   
3             0         0            0                    0                 0   
4             0         0            0                    0                 0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                   0              0          0           0           0   
1                   0              0          0           0           0   
2                   0              0          0           0           0   
3                   0              0          0           0           0   
4                   0              0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       0     0    0                 0                   0      0   
1       0     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0           0              0          0         0                     0   
1           0              0          0         0                     0   
2           0              0          0         0                     0   
3           0              0          0         0                     0   
4           0              0          0         0                     0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Building  Burger Joint  \
0               0        0                0         0             0   
1               0        0                0         0             0   
2               0        0                0         0             0   
3               0        0                0         0             0   
4               0        0                0         0             0   

   Burrito Place  Bus Line  Butcher  Café  Cajun / Creole Restaurant  \
0              0         0        0     0                          0   
1              0         0        0     0                          0   
2              0         0        0     0                          0   
3              0         0        0     0                          0   
4              0         0        0     0                          0   

   Candy Store  Caribbean Restaurant  Ch

In [46]:
# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]
toronto_onehot.rename(columns={'TorNeighborhood':'CityNeighborhood'},inplace=True)

In [47]:
print(toronto_onehot.shape)
toronto_onehot.loc[0:1700:200,:]

(1702, 231)


CityNeighborhood  Afghan Restaurant  Airport  \
0                                    The Beaches                  0        0   
200                   Cabbagetown,St. James Town                  0        0   
400                      Garden District,Ryerson                  0        0   
600                                  Berczy Park                  0        0   
800                       Adelaide,Richmond,King                  0        0   
1000     Toronto Dominion Centre,Design Exchange                  0        0   
1200     Grange Park,Kensington Market,Chinatown                  0        0   
1400       Underground city,First Canadian Place                  0        0   
1600  Parkdale Village,Brockton,Exhibition Place                  0        0   

      Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                      0             0               0                0   
200                    0             0               0                0   
400                    0             0               0                0   
600                    0             0               0                0   
800                    0             0               0                0   
1000                   0             0               0                0   
1200                   0             0               0                0   
1400                   0             0               0                0   
1600                   0             0               0                0   

      Airport Terminal  American Restaurant  Antique Shop  Aquarium  \
0                    0                    0             0         0   
200                  0                    0             0         0   
400                  0                    0             0         0   
600                  0                    0             0         0   
800                  0                    0             0         0   
1000                 0                    0             0         0   
1200                 0                    0             0         0   
1400                 0                    0             0         0   
1600                 0                    0             0         0   

      Art Gallery  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0               0                    0                 0                   0   
200             0                    0                 0                   0   
400             1                    0                 0                   0   
600             0                    0                 0                   0   
800             0                    0                 0                   0   
1000            0                    0                 0                   0   
1200            0                    0                 0                   0   
1400            0                    0                 0                   0   
1600            0                    0                 0                   0   

      Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0                 0          0           0           0       0     0    0   
200               0          0           0           0       0     0    0   
400               0          0           0           0       0     0    0   
600               0          0           0           0       0     0    0   
800               0          0           0           0       0     0    0   
1000              0          0           0           0       0     0    0   
1200              0          0           0           0       0     0    0   
1400              0          0           0           0       0     0    0   
1600              0          0           0           0       0     0    0   

      Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                    0                   0      0                0         0   
200                  0               

#### **Now let's group rows by neighborhood and by taking the means of the frequency of occurrence of each category.**

In [48]:
toronto_onehot.head()
toronto_group = toronto_onehot.groupby('CityNeighborhood').mean().reset_index()
toronto_group.loc[0:38:5,:]

CityNeighborhood  Afghan Restaurant  \
0                            Adelaide,Richmond,King                0.0   
5                                          Christie                0.0   
10               Forest Hill North,Forest Hill West                0.0   
15                     High Park,The Junction South                0.0   
20                North Midtown,Yorkville,The Annex                0.0   
25                            Roncesvalles,Parkdale                0.0   
30                  Stn A PO Boxes 25 The Esplanade                0.0   
35  Toronto Islands,Harbourfront East,Union Station                0.0   

    Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0       0.0                 0.0           0.0             0.0   
5       0.0                 0.0           0.0             0.0   
10      0.0                 0.0           0.0             0.0   
15      0.0                 0.0           0.0             0.0   
20      0.0                 0.0           0.0             0.0   
25      0.0                 0.0           0.0             0.0   
30      0.0                 0.0           0.0             0.0   
35      0.0                 0.0           0.0             0.0   

    Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0               0.0               0.0             0.030000      0.000000   
5               0.0               0.0             0.000000      0.000000   
10              0.0               0.0             0.000000      0.000000   
15              0.0               0.0             0.000000      0.000000   
20              0.0               0.0             0.045455      0.000000   
25              0.0               0.0             0.000000      0.000000   
30              0.0               0.0             0.010101      0.010101   
35              0.0               0.0             0.000000      0.000000   

    Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0       0.00     0.000000             0.000000              0.03   
5       0.00     0.000000             0.000000              0.00   
10      0.00     0.000000             0.000000              0.00   
15      0.00     0.000000             0.041667              0.00   
20      0.00     0.000000             0.000000              0.00   
25      0.00     0.000000             0.000000              0.00   
30      0.00     0.020202             0.000000              0.00   
35      0.05     0.010000             0.000000              0.00   

    Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0             0.000000            0.0   0.000000    0.000000    0.000000   
5             0.058824            0.0   0.000000    0.058824    0.000000   
10            0.000000            0.0   0.000000    0.000000    0.000000   
15            0.000000            0.0   0.000000    0.000000    0.000000   
20            0.000000            0.0   0.045455    0.000000    0.000000   
25            0.000000            0.0   0.000000    0.000000    0.000000   
30            0.000000            0.0   0.010101    0.000000    0.010101   
35            0.000000            0.0   0.000000    0.000000    0.000000   

      Bakery      Bank       Bar  Baseball Stadium  Basketball Stadium  Beach  \
0   0.030000  0.000000  0.040000              0.00            0.000000    0.0   
5   0.000000  0.000000  0.000000              0.00            0.000000    0.0   
10  0.000000  0.000000  0.000000              0.00            0.000000    0.0   
15  0.041667  0.000000  0.083333              0.00            0.000000    0.0   
20  0.000000  0.000000  0.000000              0.00            0.000000    0.0   
25  0.000000  0.066667  0.066667              0.00            0.000000    0.0   
30  0.030303  0.000000  0.000000              0.00            0.010101    0.0   
35  0.020000  0.000000  0.020000              0.02            0.010000    0.0   

    Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant    Bist

In [49]:
toronto_group.shape

(38, 231)

#### Let's print each neighborhood along with the top 5 most common venues

In [50]:
num_top_venues = 5
count=0

for hood in toronto_group['CityNeighborhood']:
    print("----"+hood+"----")
    temp = toronto_group[toronto_group['CityNeighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    count=count+1
    if (count-1)%10==0:
        print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
        print('CityNeighborhood sequence number as ',count-1,'\n')
    else:
        print('Note here only print every 10th result to check',' \n')

----Adelaide,Richmond,King----
             venue  freq
0      Coffee Shop  0.07
1             Café  0.05
2              Bar  0.04
3       Steakhouse  0.04
4  Thai Restaurant  0.03
CityNeighborhood sequence number as  0 

----Berczy Park----
Note here only print every 10th result to check  

----Business Reply Mail Processing Centre 969 Eastern----
Note here only print every 10th result to check  

----Cabbagetown,St. James Town----
Note here only print every 10th result to check  

----Central Bay Street----
Note here only print every 10th result to check  

----Christie----
Note here only print every 10th result to check  

----Church and Wellesley----
Note here only print every 10th result to check  

----Davisville----
Note here only print every 10th result to check  

----Davisville North----
Note here only print every 10th result to check  

----Dufferin,Dovercourt Village----
Note here only print every 10th result to check  

----Forest Hill North,Forest Hill West----
          

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [51]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [52]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['CityNeighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['CityNeighborhood'] = toronto_group['CityNeighborhood']

for ind in np.arange(toronto_group.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_group.iloc[ind, :], num_top_venues)

print('Dataframe shape is',neighborhoods_venues_sorted.shape)
neighborhoods_venues_sorted.loc[0:38:5,:]

Dataframe shape is (38, 11)


CityNeighborhood 1st Most Common Venue  \
0                            Adelaide,Richmond,King           Coffee Shop   
5                                          Christie         Grocery Store   
10               Forest Hill North,Forest Hill West                  Park   
15                     High Park,The Junction South       Thai Restaurant   
20                North Midtown,Yorkville,The Annex                  Café   
25                            Roncesvalles,Parkdale             Gift Shop   
30                  Stn A PO Boxes 25 The Esplanade           Coffee Shop   
35  Toronto Islands,Harbourfront East,Union Station           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Café                   Bar            Steakhouse   
5                   Café                  Park            Restaurant   
10         Jewelry Store                 Trail      Sushi Restaurant   
15                   Bar                  Café    Mexican Restaurant   
20        Sandwich Place           Coffee Shop           Pizza Place   
25           Coffee Shop             Bookstore                  Bank   
30            Restaurant                  Café    Seafood Restaurant   
35                 Hotel              Aquarium                  Café   

     5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Breakfast Spot      Asian Restaurant        Cosmetics Shop   
5       Italian Restaurant                 Diner            Baby Store   
10             Yoga Studio                 Diner    Falafel Restaurant   
15                    Park        Discount Store                 Diner   
20  Furniture / Home Store              Pharmacy     Convenience Store   
25      Italian Restaurant         Movie Theater               Dog Run   
30                Beer Bar                 Hotel   Japanese Restaurant   
35     Fried Chicken Joint    Italian Restaurant            Restaurant   

   8th Most Common Venue      9th Most Common Venue  \
0                  Hotel                 Restaurant   
5     Athletics & Sports                  Nightclub   
10           Event Space       Ethiopian Restaurant   
15                Bakery  Cajun / Creole Restaurant   
20        Cosmetics Shop                        Pub   
25            Restaurant               Dessert Shop   
30    Italian Restaurant                     Bakery   
35               Brewery             Scenic Lookout   

         10th Most Common Venue  
0           American Restaurant  
5                   Candy Store  
10            Electronics Store  
15          Fried Chicken Joint  
20                 Burger Joint  
25  Eastern European Restaurant  
30                       Lounge  
35                  Music Venue

## Now run K-Means Clustering to cluster the neighborhood into 8 clusters.

In [53]:
# set number of clusters
kclusters = 10

toronto_group_clustering = toronto_group.drop('CityNeighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_group_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 6, 1, 1, 1, 1, 1, 8, 7], dtype=int32)

In [57]:
neighborhoods_venues_sorted.drop(columns='Cluster Labels',inplace=True)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [58]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = table_tor

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('CityNeighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

PostalCode          Borough                   Neighborhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    Riverdale,The Danforth West  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               9          Neighborhood                   Pub   
1 -79.352188               1      Greek Restaurant           Coffee Shop   
2 -79.315572               1                  Park            Board Shop   
3 -79.340923               1                  Café           Coffee Shop   
4 -79.388790               3                  Park           Swim School   

  3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
0                  Park     Health Food Store                   Trail   
1        Ice Cream Shop    Italian Restaurant  Furniture / Home Store   
2                   Pub               Brewery            Burger Joint   
3                Bakery    Italian Restaurant     American Restaurant   
4              Bus Line           Yoga Studio                   Diner   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0    Dim Sum Restaurant                 Diner        Discount Store   
1          Liquor Store                   Spa             Juice Bar   
2         Burrito Place  Fast Food Restaurant        Sandwich Place   
3        Sandwich Place      Stationery Store           Fish Market   
4    Falafel Restaurant           Event Space  Ethiopian Restaurant   

  9th Most Common Venue       10th Most Common Venue  
0               Dog Run             Doner Restaurant  
1             Bookstore                      Brewery  
2     Fish & Chips Shop                   Steakhouse  
3       Coworking Space    Latin American Restaurant  
4     Electronics Store  Eastern European Restaurant

#### Finally let's visualize the resulting clusters.

In [56]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=10,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-2],
        fill_opacity=2).add_to(map_clusters)
       
map_clusters
#map_toronto